Dev/break connection

qnikst opened the pull request.

qnikst committed:
Expose internal state and debugging API

qnikst commented:
This PR solves issue haskell-distributed/network-transport-inmemory#9 and affect haskell-distributed/distributed-process#419.

mboes commented on src/Network/Transport/InMemory.hs:
```diff
@@ -2,7 +2,12 @@
+module Network.Transport.InMemory
+  ( createTransport
+  , createTransportExposeInternals
+  -- * Debug
```
say "For testing purposes". The function should not be called `apiBreakConnection` which is an internal name. Introduce a wrapper for externally exporting this function as `breakConnection`.

mboes replied:
dropped?

mboes commented:
:+1: apart from my minor comments above.

qnikst commented:
I've addressed comments, but introduced two more changes:
1. expose `newtype TransportInternals` instead of `TVar TransportState` without exporting constructor
2. changed `breakConnection` live in `IO` instead of `STM`

mboes commented on src/Network/Transport/InMemory.hs:
```diff
@@ -62,17 +67,26 @@ data LocalConnectionState
   | LocalConnectionClosed
   | LocalConnectionFailed
 
+newtype TransportInternals = TransportInternals (TVar TransportState)
```
Needs to be exported too. Otherwise expose internals useless. Needs a docstring. To the effect of "Internal state of the transport."

qnikst replied:
I thought that user could still use that datatype. Will fix now

mboes commented on src/Network/Transport/InMemory.hs:
```diff
@@ -62,17 +67,26 @@ data LocalConnectionState
+
+-- | Create a new Transport exposing internal state.
+--
+-- This state can be used for with internal or debug functions, e.g.
```
"Useful for testing and/or debugging purposes. Should not be used in production. No guarantee as to the stability of the internals API."

mboes commented:
remember to address comment about haddock section in export list.

qnikst commented:
Not sure that I understood `dropped` comment about export list, is it about `TransportInternals (..)`?

mboes commented:
no expand all "outdated diffs" above. You'll see.

qnikst commented:
I see, fixed in haskell-distributed/network-transport-inmemory#12 .

mboes commented:
No, not fixed in haskell-distributed/network-transport-inmemory#12. Copying here:

> say "For testing purposes". The function should not be called apiBreakConnection which is an internal name. Introduce a wrapper for externally exporting this function as breakConnection.

In reference to the section name currently called "Debug". If I could push code right now I'd just do it myself rather than this back and forth.

qnikst commented:
I've opened haskell-distributed/network-transport-inmemory#13, I hope that PR will work reliably. For some reason in haskell-distributed/network-transport-inmemory#12 github showed outdated version in "changed files"..
